File: src/utils/poa/publisher.js
import {shortHash, showNotification, regexAddress} from '../../ui/Utils'

let crypto = require('crypto-browserify')
let jsrsasign = require('jsrsasign')

const POA_PROTOCOL_VERSION = 4

const methods = {
    'ERR_DUPLICATE_KEY':'ERR_DUPLICATE_KEY',
    'on_leader_beacon':'on_leader_beacon',
    'peer':'peer',
}

class Publisher {


    constructor(account, token, connector = false, ip = "95.216.246.116", port = "3000") {

        // this

        // token = '0d0b6da9d730e6eae5e6cff889051d909b8f66be0f4dc315c3fcedf27395c0cb'
        // token = 'de0942b3b1194cde66ba9bf45bd1bdf406e714d6d514b8c0e6fd58b5ee833693'
        console.log(token)

        let id = account.publicKey.slice(0, 6)
        this.connector = connector
        this.restart = true
        this.ws = new WebSocket(`ws://${ip}:${port}`)
        this.account = account
        this.token = token
        this.ip = ip
        this.port = port
        this.status = 'Initialisation'

        this.close = () => {
            this.restart = false
            this.status = 'Closed'
            console.log(`${id} closed`)
            this.ws.close()
        }

        this.init = init

        this.init(this, id, this.ip, this.port, this.account, this.token)
    }
}

function init(_, id, ip, port, account, token) {
    _.ws.onopen = () => {
        _.restart = true
        console.log(`${id} connected`)
        let hash = crypto.createHash('sha256')
            .update(ip)
            .digest('hex')

        let hail = {
            'data': {
                'hash': hash,
                'id': account.publicKey,
                'token': token,
                'sign': sign(account.privateKey, hash)
            },
            'method': 'hail',
            'ver': POA_PROTOCOL_VERSION
        }
        if (account.hasOwnProperty('referrer') && regexAddress.test(account.referrer)) {
            hail.data.referrer = account.referrer
        }
        _.ws.send(JSON.stringify(hail))

        _.status = 'Connected'
    }

    _.ws.onclose = (e) => {
        console.log(`${id} disconnected`)
        if (_.restart) {
            setTimeout(() => {
                console.log(`${id} restarted`)
                _.ws = new WebSocket(`ws://${_.ip}:${_.port}`)
                init(_, id, _.ip, _.port, account, token)
            }, 5000)
        } else {
            _.status = 'Disconnected'
        }
    }

    _.ws.onerror = (err) => {
        console.error('Socket encountered error: ', err.message, 'Closing socket')
        // init(_, id, ip, account, token)
        _.status = 'Error'
        _.close()
    }
    _.ws.onmessage = (msg) => {
        try {
            msg = JSON.parse(msg.data)
        } catch (er) {
            console.error(er)
            return
        }


        if (msg.err === 'ERR_DUPLICATE_KEY') {
            console.warn(msg)
            _.status = 'ERR_DUPLICATE_KEY'
            // if(_.connector != false){
            //     _.connector(JSON.stringify({ method: 'notification', body:{title:"Mining error",text:'Duplicate key by ' + shortHash(account.publicKey)}}))
            // }else{
            //     showNotification('Mining error', 'Duplicate key by ' + shortHash(account.publicKey))
            // }
        }

        if (methods[msg.method] === undefined) {
            _.status = 'Unknown method'
            return
        }

        // if(msg.method === 'peer'){
        //     _.ip = msg.data.ip
        //     _.port = msg.data.port
        //     _.status = "Peer change"
        //     _.close()
        // }

        if(msg.method === methods.on_leader_beacon){
            let data = msg.data
            let isValid = true
            let isCorrect = (hashBlock(data.mblock_data) === data.m_hash)
    
            console.log(` ${id}  Sign: ${(isValid ? 'OK' : 'BAD')}  m_hash: ${data.m_hash}  ${(isCorrect ? 'OK' : 'BAD')}`)
            //console.log(`poaId: ${poa.id}-${i}   Sign: ${(isValid?"OK":"BAD")}   ${(isCorrect?"OK":"BAD")}`);
    
            if (!isValid) {
                //console.log("Incorrect sign")
                _.status = 'Incorrect sign'
                return
            }
            if (!isCorrect) {
                //console.log("Incorrect m_hash")
                return
            }
    
            // let token = tokens[Math.random() >= 0.8 ? 1 : 0]
            let forSign = data.m_hash + (account.hasOwnProperty('referrer') && regexAddress.test(account.referrer) ? account.referrer : '') + token
    
            let res = {
                'ver': POA_PROTOCOL_VERSION,
                'method': 'publish',
                'data': {
                    'kblocks_hash': data.mblock_data.kblocks_hash,
                    'm_hash': data.m_hash,
                    'token': token,
                    'sign': sign(account.privateKey, forSign),
                    'id': account.publicKey
                }
            }
    
            if (account.hasOwnProperty('referrer') && regexAddress.test(account.referrer)) {
                res.data.referrer = account.referrer
            }
    
            _.status = 'Sign block'
    
            // if(_.connector != false){
            //     _.connector(JSON.stringify({ method: 'notification', body:{title:'Mining reward',text:'Sign block by ' + shortHash(account.publicKey)}}))
            // }else {
            //     showNotification('Mining reward', 'Sign block by ' + shortHash(account.publicKey))
            // }
    
            _.ws.send(JSON.stringify(res))
        }   
    }
}

function sign(privateKey, msg) {
    var sig = new jsrsasign.Signature({'alg': 'SHA256withECDSA'})
    sig.init({
        d: privateKey,
        curve: 'secp256k1'
    })
    sig.updateString(msg)

    return sig.sign()
}

function verify(publicKey, msg, signedMsg) {
    var sig = new jsrsasign.Signature({'alg': 'SHA256withECDSA'})
    sig.init({
        xy: publicKey,
        curve: 'secp256k1'
    })
    sig.updateString(msg)
    return sig.verify(signedMsg)
}

function hashBlock(block) {
    let txs_hash = crypto.createHash('sha256')
        .update(block.txs.map(tx => hash_tx(tx))
            .sort()
            .join(''))
        .digest('hex')

    let hashed = crypto.createHash('sha256')
        .update(block.kblocks_hash.toLowerCase() + block.nonce.toString() + block.publisher.toLowerCase() + txs_hash.toLowerCase())
        .digest('hex')
    return hashed
}

function hash_tx(tx) {
    if (!tx) {
        return undefined
    }

    let str = ['amount', 'data', 'from', 'nonce', 'sign', 'ticker', 'to'].map(v => crypto.createHash('sha256')
        .update(tx[v].toString()
            .toLowerCase())
        .digest('hex'))
        .join('')

    return crypto.createHash('sha256')
        .update(str)
        .digest('hex')
}


export {Publisher}

File: android/app/src/main/java/com/enecuum/pwa/Publisher.java
package com.enecuum.pwa;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.net.URI;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URISyntaxException;

import com.getcapacitor.JSObject;
import com.google.gson.Gson;

public class Publisher {

    private String TAG = "Publisher";
    private final String privateKey;
    public String token;
    public String referrer;
    public URI uri;
    public String publicKey;
    public Crypto crypto = new Crypto();
    public String ip;
    public String port;
    public String wsUrl;
    public Integer protocol_version = 4;

    public Boolean restartMiner;
    public String status = "Disconnected";
    public Boolean mining;
    public Integer countBlocks = 0;
    public Boolean reboot;

    public WebSocketClient ws;


    Publisher(String url, String port, String privateKey, String token, String referrer) {
        this.privateKey = privateKey;
        this.publicKey = crypto.getPublicKey(this.privateKey);
        this.token = token;
        this.referrer = referrer;
        this.ip = url;
        this.port = port;
        this.wsUrl = String.format("ws://%s:%s/", url, port);
        this.reboot = false;

        try {
            uri = new URI(wsUrl);

            ws = new WebSocketClient(this.uri) {
                @Override
                public void onOpen(ServerHandshake handshakeData) {
                    try {
                        Log.d(TAG, "connect from " + publicKey.substring(0, 6) + " to " + this.uri.toString());
                        ws.send(hail());
                        status = "Connected";
                    } catch (Exception ex) {
                        Log.d(TAG, "Error in open WS\n" + ex.getMessage() + "\n" + ex.getStackTrace());
                        reboot = true;
                        status = "Disconnected";
                    }

                }

                @Override
                public void onMessage(String message) {
                    Log.d(TAG, "account " + publicKey.substring(0, 6) + " take block");
                    try {
                        String answer = onBlock(message);
                        if (answer.length() > 0) {
                            ws.send(answer);
                        }
                    } catch (Exception ex) {
                        Log.d(TAG, "Error in block\n" + ex.getMessage() + "\n" + ex.getStackTrace());
                        status = "Disconnected";
                        ws.close();
                        reboot = true;
                    }
                }

                @Override
                public void onClose(int code, String reason, boolean remote) {
                    try {
                        ws.close();
                    } catch (Exception ex) {
                        Log.d(TAG, "WS is closed");
                    }
                    if (restartMiner) {
                        reboot = true;
                    }
                }

                @Override
                public void onError(Exception ex) {
                    Log.d(TAG, "ERROR: " + ex.getMessage() + "\n" + ex.getStackTrace());
                    try {
                        ws.close();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    if (restartMiner) {
                        reboot = true;
                    }
                }
            };

        } catch (URISyntaxException e) {
            Log.d(TAG, "uri error");
            e.printStackTrace();
            reboot = true;
        }
//        System.out.println(this.wsUrl);
//        System.out.println(this.uri);


    }

    public void init() {
        try {
            if (this.mining) {
                restartMiner = true;
                countBlocks = 0;
                ws.connect();
            } else {
                Log.d(TAG, "Miner " + this.publicKey.substring(0, 6) + " deactivated");
                this.status = "Disconnected";
//                this.reboot = true;
            }

        } catch (Exception ex) {
            Log.d(TAG, "Error ws: " + ex.getMessage() + "\n" + ex.getStackTrace());
            this.status = "Disconnected";
            this.reboot = true;
        }
    }

    public void stop() {
        try {
            restartMiner = false;
            ws.close();
        } catch (Exception ex) {
            Log.d(TAG, "Error ws: " + ex.getMessage() + "\n" + ex.getStackTrace());
        }
        this.status = "Disconnected";
    }

    public String hail() {
        String hash = crypto.sha256(this.ip + (this.referrer.length() > 0 ? this.referrer : "") + this.token);
//        String hash = crypto.sha256(this.ip + this.token);
        JSObject obj = new JSObject();
        JSObject hail = new JSObject();
        hail.put("hash", hash);
        hail.put("id", this.publicKey);
        hail.put("token", this.token);
        hail.put("sign", crypto.sign(this.privateKey, hash));
        if (this.referrer.length() > 0) {
            hail.put("referrer", this.referrer);
        }
        obj.put("data", hail);
        obj.put("method", "hail");
        obj.put("ver", protocol_version);

        Log.d(TAG, obj.toString());
        return obj.toString();
    }


    public String onBlock(String m_block) {
        Gson g = new Gson();
        Block block = g.fromJson(m_block, Block.class);
        Log.d(TAG, block.method);

        if (block.err != null) {
            if (block.err.equals("ERR_DUPLICATE_KEY")) {
                Log.d(TAG, "ERR_DUPLICATE_KEY!");
                reboot = true;
                return "";
            }
        }

        if (block.method.equals("peer")) {
            this.stop();
            Peer peer = g.fromJson(m_block, Peer.class);
            IPUpdater ipUpdater = new IPUpdater();
            ipUpdater.publicKey = this.publicKey;
            ipUpdater.ip = peer.data.ip;
            ipUpdater.port = peer.data.port;
            String send = ipUpdater.serialize();
            PoA.ipUpdaterList.add(send);
        }

        if (block.method.equals("on_leader_beacon")) {
            String msg = block.data.m_hash + (this.referrer.length() > 0 ? this.referrer : "") + this.token;
            JSObject obj = new JSObject();
            JSObject publish = new JSObject();
            publish.put("kblocks_hash", block.data.mblock_data.kblocks_hash);
            publish.put("m_hash", block.data.m_hash);
            publish.put("id", this.publicKey);
            publish.put("token", this.token);
            if (this.referrer.length() > 0) {
                publish.put("referrer", this.referrer);
            }
            publish.put("sign", crypto.sign(this.privateKey, msg));
            obj.put("data", publish);
            obj.put("method", "publish");
            obj.put("ver", protocol_version);

            this.countBlocks++;
//            this.status = "Sign block";
            this.status = String.format("Sign block (%d)", countBlocks);

            Log.d(TAG, obj.toString());
            return obj.toString();
        }
        Log.d(TAG, "not found");
        return "";
    }
}
